feat: allow caching and resuming picker

fdschmidt93 opened the pull request:
Will close #751

What?
- Introduce `resume` (resume last picker) and `pickers` (picker over cached pickers) `builtin` pickers

How?
- Modify `pickers.lua`  so picker attributes are not `nil`-ed on detach but stored in global `cached_pickers` table-stack instead
- `cache_limit` > 0 reduces # of cached entries via `LinkedList:truncate` (primarily to reduce size)
- New buffer previewer that writes original results in preview

Showcase

https://user-images.githubusercontent.com/39233597/127784598-e3d32413-2957-4fc5-8a1e-8ea6598b42b2.mp4

fdschmidt93 commented on lua/telescope/pickers.lua:
```diff
@@ -401,18 +405,31 @@ function Picker:find()
+        end
+      else
+        -- resume previous picker
+        local index = 1
+        for entry in self.manager:iter() do
+          self:entry_adder(index, entry, _, true)
+          index = index + 1
+        end
+        self.cached_picker = false
+        -- TODO(fdschmidt93) this does not work for live_grep
+        -- redraw highlighting for multi-selections
+        self.refresh()
```
I've been dreading already too long over this.

The straightforward approach of 

```lua
local index = 1
for entry in self.manager:iter() do
  self:entry_adder(index, entry, _, true)
  index = index + 1
  if self:can_select_row(row) and self._multi:is_selected(entry) then
    self.highlighter:hi_multiselect(row, true)
  end
end
```

doesn't work. I've tried a lot of other hacks I could think of but to no avail. In any case, selected entries in any case would be highlighted once you scroll over them.

fdschmidt93 commented on lua/telescope/pickers.lua:
```diff
@@ -401,18 +405,31 @@ function Picker:find()
+          log.warn("Failed with msg: ", msg)
+        end
+      else
+        -- resume previous picker
+        local index = 1
+        for entry in self.manager:iter() do
+          self:entry_adder(index, entry, _, true)
+          index = index + 1
+        end
+        self.cached_picker = false
+        -- TODO(fdschmidt93) this does not work for live_grep
+        -- redraw highlighting for multi-selections
+        self:refresh()
```
I've been dreading already too long over this.

The straightforward approach of 

```lua
local index = 1
for entry in self.manager:iter() do
  self:entry_adder(index, entry, _, true)
  index = index + 1
  if self:can_select_row(row) and self._multi:is_selected(entry) then
    self.highlighter:hi_multiselect(row, true)
  end
end
```

doesn't work. I've tried a lot of other hacks I could think of but to no avail. In any case, selected entries in any case would be highlighted once you scroll over them.

tjdevries commented on lua/telescope/algos/linked_list.lua:
```diff
@@ -215,4 +215,25 @@ function LinkedList:ipairs()
   end
 end
 
+function LinkedList:truncate(max_results)
```
Hmm, we could decide whether to go from the front or the end based on the length and the max results.

For example, if we have 500,000 results and we limit 700,000 then there's actually no reason to go at all.

If we have 1,000,001 results and limit is 1,000,000 better to just truncate the last entry.

tjdevries commented on doc/telescope.txt:
```diff
@@ -23,6 +23,25 @@ telescope.setup({opts})                                    *telescope.setup()*
+        Only relevant if `cache_picker = true`. Set to -1 to avoid
+        caching limit.
+
+        Default: 10000
```
I think our default could be even smaller if it affects performance at all.

fdschmidt93 replied:
Agreed

fdschmidt93 commented:
I'd be to get advice on whether the buffer previewer can be done more cleverly to universally account for (filtered) ascending or descending results. Lines apparently cannot be placed at an arbitrary row for empty buffers.

tami5 commented:
> It would be great to persist pickers across sessions by writing them to disk as we then can preserve multi selections for cool stuff.
> 
> 
> 
> However, it quite challenging:
> 
> - Pickers across projects are mostly irrelevant to one another -- requires likely something like git root resolved workspace storage
> 
> - Serializing lua tables appears tedious for that use case ... really happy to hear ideas on this as I think it's super cool :)

I'd say use sql.nvim, it is fully equipped with this kind of stuff 🤣. But @tjdevries hate being tide to yet another dependency. Alternatively, look at how I serialized and deserialized lua tables, it's simply. then maybe save and load then from a file.

btw great job @fdschmidt93

tjdevries commented:
@fdschmidt93 I think saving to disk would make a good extension :) (as @tami5 would predict that I would say). If we end up finding simple, OS agnostic, performant solutions over time there, then we can talk about moving it back into core, but I do not want to hold up a PR and/or create undue complexity for that feature at this time.

Additionally, if you make it an extension, you can require sql.nvim, or other luarocks packages if desired.

tjdevries commented:
I do not like that the pickers are cached forever. We should make a maximum size so that we release the pickers at some point in the future (for example, another config option is OK to do perhaps). We should set this to something small by default, perhaps only 1.

This will prevent keeping around 100,000 entries for the rest of your nvim editing session, which would be very confusing for people who just boot up telescope for the first time and live grep somewhere a few times.

tjdevries commented:
Also, if we add more entries, then I would move us towards doing something like having a `resume` option that takes several configuration options for pickers. Idk, I'd have to think a bit more about that generally then.

fdschmidt93 commented:
Thanks @tami5 :)

I've briefly looked at `sql.nvim/json.lua` and if I understand all correctly it's not look I cannot easily dump a complex table like a picker and its finders, entries, and selections in there without significant wrappers or telescope.nvim changes to make my life easier. As in, the way to probably best use `sql.nvim` would be to parse entries, selections, and whatever required to reinstantiate the picker and then run `find`. Coming from Python, I was looking for something like `pickle` :sweat_smile: 

@tjdevries 

Agreed on the extension :)

Makes sense, I'll add a limit of # cached pickers and expose it as a config. For context, as is, it's not really an issue since the entries are truncated to the first 1K by default.

tjdevries commented:
I can't really imagine a usecase for looking at the 327th picker ago :laughing:

fdschmidt93 commented:
We now
- Cache only `cache_n_pickers` (defaullt: 10, @tjdevries if you feel strongly about only the last, happy to change though from all my tests it should be fine)
- Can override `opts` of `builtin.resume` and `builtin.pickers` (latter passed onto `resume` ofc) consistent with other pickers

:)

Conni2461 commented on lua/telescope/builtin/internal.lua:
```diff
@@ -1,12 +1,15 @@
 local actions = require "telescope.actions"
 local action_set = require "telescope.actions.set"
 local action_state = require "telescope.actions.state"
+local config = require "telescope.config"
```
Not needed we already have `conf` imported (unless you need something else from config, doesn't look like it rn)

fdschmidt93 replied:
Yes, slight oversight sorry :) fixed locally and to be updated once more comes in

Conni2461 commented on lua/telescope/config.lua:
```diff
@@ -279,6 +279,42 @@ local telescope_defaults = {
   ]],
   },
 
+  cache_picker = {
```
Rather than having 3 (or more) values could we do something similar to `layout_config` and `history` where we have one table setting with multiple keys?

fdschmidt93 replied:
I hope this is fine now; I'm afraid this increased control flow complexity quite a bit as I'm writing to the `pickers.cache_picker` during caching and `opts.cache_picker = false` can be validly passed to `builtin.{resume, pickers}`.

Conni2461 replied:
I like the config now. I am more concerned about a better user interface than a couple more ifs in code :laughing: 
Also i love this table config to group stuff since clason proposed it for history. way better than having multiple string values that are prefixed with something

Conni2461 commented on lua/telescope/pickers.lua:
```diff
@@ -490,6 +529,8 @@ function Picker:find()
 
   if self.initial_mode == "insert" then
     vim.cmd [[startinsert!]]
+    -- make starting in insert mode more robust
+    vim.cmd [[normal! A]]
```
Is this a workaround for opening a picker in an action if telescope wasn't closed with `actions._close(prompt_bufnr, true)`

fdschmidt93 replied:
Sort of, yes 😅 as I didn't recall `_close` vs `close` in `builtin.internal.pickers`.

Not sure whether to keep this (since it shouldn't hurt?) or only change to `_close` in pickers action.

Conni2461 commented on lua/telescope/builtin/internal.lua:
```diff
@@ -71,6 +74,56 @@ internal.builtin = function(opts)
+      actions.select_default:replace(function(prompt_bufnr)
+        local current_picker = action_state.get_current_picker(prompt_bufnr)
+        local selection_index = current_picker:get_index(current_picker:get_selection_row())
+        actions.close(prompt_bufnr)
```
You should close with `actions._close(prompt_bufnr, true)`

Conni2461 commented on lua/telescope/previewers/buffer_previewer.lua:
```diff
@@ -858,6 +860,77 @@ previewers.highlights = defaulter(function(_)
+      putils.with_preview_window(status, nil, function()
+        local ns_telescope_entry = vim.api.nvim_create_namespace "telescope_entry"
+        local preview_height = vim.api.nvim_win_get_height(status.preview_win)
+        if not self.state.bufname then
```
Can we flip this to make it a little bit more readable. `if self.state.bufname then return end`. I currently having a hard time with the indents and all the ends :laughing:

sodiumjoe commented:
Trying out this branch, a couple of notes few notes:
* it would be great if it would also resume the selection state
* it would be great if it would also resume the input buffer mode

The workflow i'm thinking of is quickfix-ish, i.e. I grep for something, then go through the list of results making edits. So when I resume, I want to be in normal mode, with the same selection highlighted.

fdschmidt93 commented:
Many thanks for trying out the branch and the valuable feedback :) I've incorporated your suggestions as far as possible right now.

There still are a couple of issues around initial normal mode, most prominently wrt caching `default_text` does not play nicely and results are only shown once insert mode was entered. I guess most, if not all, issues around this will be fixed in #987 

I'll most likely have to rebase this branch then to #987 anyways and will validate then that these cases around caching work as expected.

Thanks again!

sodiumjoe commented:
Thanks, this is great! Now that I think about it, I reported https://github.com/nvim-telescope/telescope.nvim/issues/750 which might have something to do with the issues starting in normal mode. I'll continue testing this branch, but so far so good!

tjdevries commented on lua/telescope/config.lua:
```diff
@@ -279,6 +290,33 @@ local telescope_defaults = {
 
+  cache_picker = {
+    {
+      num_pickers = 20,
```
I really think this should be 1 by default. If users are interested in storing lots of em, they can do that.

Otherwise, this lets them do the most common use case, which is resuming the last thing they did. I don't think this will happen often to store a bunch of them.

tjdevries commented on lua/telescope/pickers.lua:
```diff
@@ -401,18 +408,53 @@ function Picker:find()
+        if not ok then
+          log.warn("Failed with msg: ", msg)
+        end
+      else
```
Hmmm, this section makes me a bit nervous.

We have a few vim.schedules, which I'm hoping we should not have to do necessarily.

Also, with newest update (you'll have to pull from master again) you should be able to do something like await_scheduler() instead, if you really need to do scheduling to make this work out OK.

Also, I'd see if we can re-use more code here or abstract it into some helper function within picker so that we don't share too much state here accidentally.

Does that make sense?

fdschmidt93 replied:
> re-use more code here

The function for highlighting was already there :laughing: I just hadn't seen it.

Yes, `await_scheduler` is required to make this work.

tjdevries commented on lua/telescope/previewers/buffer_previewer.lua:
```diff
@@ -234,6 +234,8 @@ previewers.new_buffer_previewer = function(opts)
       end
     end
+    -- enable resuming picker with existing previewer to avoid lookup of deleted bufs
+    bufname_table = {}
```
I can't find anywhere that this is being used (on first search, just in github browser)

Conni2461 replied:
getting a new table ref here should be good. I think we can ignore that comment :laughing:

tjdevries commented:
Other than a few nitpicks, I just want to make sure that this still works with the newest async updates and maybe it will allow writing some of the code a bit easier and more straightforward in the pickers.lua section.

Once we do that, I think we can probably merge!

fdschmidt93 commented:
I've rebased to latest master and addressed the comments. One thing that doesn't work (should it? also seems to affect master), is setting `default_text` for a picker if `initial_mode="normal"`, such that the `default_text` is shown in the prompt.

@sodiumjoe given your experience with this branch, could you also please retest it and report whether everything fine is on your end? If you are using the `pickers` picker then please set caching in your set up adequately, like so:

```lua
telescope.setup {
  defaults = {
    cache_picker = {
      num_pickers = 20 -- your preferred number here, values up to 100 should be perfectly fine; likely even much higher
    }
  }
}
```

Conni2461 commented on doc/telescope.txt:
```diff
@@ -25,6 +25,28 @@ telescope.setup({opts})                                    *telescope.setup()*
+                              If passed to a picker, the cached pickers with
+                              indices larger than `cache_picker.num_pickers` will
+                              be cleared.
+                              Default: 20
```
docs no longer align.

fdschmidt93 replied:
Sorry & thanks, fixed.

Conni2461 reviewed (approved):
LGTM

Thanks :) and good job

fdschmidt93 commented:
Thanks :)

max397574 commented:
afaict it isn't possible but ig you could opena feature request issue for this
it sounds like something really useful @yyy33

yyy33 commented:
> afaict it isn't possible but ig you could opena feature request issue for this it sounds like something really useful @yyy33

thank's , i opened it
